packages/next/server/dev/next-dev-server.ts

```diff
@@ -404,34 +457,69 @@ export default class DevServer extends Server {
               this.customRoutes.rewrites.beforeFiles.length > 0 ||
               this.customRoutes.rewrites.fallback.length > 0
 
-            config.plugins?.forEach((plugin: any) => {
-              // we look for the DefinePlugin definitions so we can
-              // update them on the active compilers
-              if (
-                plugin &&
-                typeof plugin.definitions === 'object' &&
-                plugin.definitions.__NEXT_DEFINE_ENV
-              ) {
-                const newDefine = getDefineEnv({
-                  dev: true,
-                  config: this.nextConfig,
-                  distDir: this.distDir,
-                  isClient,
-                  hasRewrites,
-                  hasReactRoot: this.hotReloader?.hasReactRoot,
-                  isNodeServer,
-                  isEdgeServer,
-                  hasServerComponents: this.hotReloader?.hasServerComponents,
-                })
-
-                Object.keys(plugin.definitions).forEach((key) => {
-                  if (!(key in newDefine)) {
-                    delete plugin.definitions[key]
+            if (tsconfigChange) {
+              config.resolve?.plugins?.forEach((plugin: any) => {
```

Comment:
I think we could actually watch / reload the compilation in the jsconfigPathsPlugin instead which would allow for triggering a rebuild as well 🤔

Reply:
If we handle this at the plugin level I think we may have extra invalidations, consolidating the invalidation in the watcher allows us to combine multiple changes into one invalidate e.g. if `.env` is changed and `tsconfig.json` in one aggregate we can apply the configs and then trigger a single invalidate.